src/addon/addonmanager.js

```diff
@@ -6,134 +6,281 @@
+            logger.debug('monitorBranch [' + branchName + '] - starting new monitor');
+            monitor.instance.start()
+                .then(deferred.resolve)
+                .catch(deferred.reject);
+        }
 
-        if (!addOn) {
-            deferred.reject(new Error('The addOn is not running'));
-        } else {
-            addOn.query(parameters, function (err, message) {
-                if (err) {
-                    deferred.reject(err instanceof Error ? err : new Error(err));
+        if (monitor) {
+            monitor.connectionCnt += 1;
+            logger.debug('monitorBranch [' + branchName + '] - connection counter [prev, now]',
+                monitor.connectionCnt - 1, monitor.connectionCnt);
+            if (monitor.connectionCnt === 1) {
+                // The client disconnected before the monitor itself connected.
+                // Register on start deferred and stop the monitor.
+                logger.debug('The client disconnected before the monitor itself trigger connect.');
+                monitor.instance.start()
+                    .then(function () {
+                        return monitor.instance.stop();
+                    })
+                    .then(function () {
+                        removeMonitor(branchName);
+                        deferred.resolve();
+                    })
+                    .catch(deferred.reject);
+            } else if (monitor.connectionCnt === 2) {
+                if (monitor.stopTimeout) {
+                    // The monitor is in stopping stage and
+                    // the timeout had not been triggered yet.
+                    clearTimeout(monitor.stopTimeout.id);
+                    logger.debug('monitorBranch [' + branchName + '] - setTimeout cleared');
+                    monitor.stopTimeout.deferred.resolve(monitor.connectionCnt);
+                    monitor.stopTimeout = null;
+                    monitor.instance.start()
+                        .then(deferred.resolve)
+                        .catch(deferred.reject);
+                } else if (monitor.instance.stopRequested === true) {
+                    // [Rare case]
+                    // The monitor is in stopping stage and the timeout has been triggered.
+                    // We cannot simply remove the monitor before it has closed the branch,
+                    // therefore we need to register on the stop.promise and start a new monitor
+                    // once it has resolved.
+                    //
+                    // (For bookkeeping of the addOnManagers it is not the same since they create a new storage.)
+                    self.inStoppedAndStarted[branchName] = true;
+                    logger.debug('monitorBranch [' + branchName + '] - inStoppedAndStarted!');
+                    monitor.instance.stop()
+                        .then(function () {
+                            delete self.inStoppedAndStarted[branchName];
```

Comment:
It does cover for the case where one new socket connects in that small time window. There is a test case for it.
But yes, multiple incoming ones should register on the stop-promise too and when that is resolved register on the start-promise. (With multiple registrations on the promise - the will resolve/reject in the order they were registered).